### PR TITLE
#163383192 Add functionality to filter and search for articles

### PR DESCRIPTION
**Pull request by oma0256**
**What does this PR do?**
 - Allow a user filter for articles based on article title, tags and author username.
 - Allow user search using keywords.

**How should this be manually tested?**
 - Download and install postman onto you're local machine
 - Open postman
- Then visit the route `https://ah-backend-summer-stagin-pr-37.herokuapp.com/api/v1/articles/?<param>=<search_text>` `<param> can be author, title, tag <search_text> the text you want to search or filter basing on` to filter for articles or `https://ah-backend-summer-stagin-pr-37.herokuapp.com/api/v1/articles/?search=<search_text>` to search for articles using `GET`.

**What are the relevant pivotal tracker stories?**
[#163383192](https://www.pivotaltracker.com/n/projects/2240000/stories/163383192)

### CODE REVIEW

File: authors/apps/articles/tests/test_filter_search_functionality.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        username = self.user.username
+        response = self.client.get(f"{self.url}?author={username}")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 1)
+        response = self.client.get(f"{self.url}?author=vdhfvsf")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 0)
+
+    def test_user_can_filter_by_articles_by_title(self):
```
Comment (fahdjamy) [inline authors/apps/articles/tests/test_filter_search_functionality.py]: Thanks for the good work done @oma0256. But think you want to do here is name this method as test_user_can_filter_articles_by_title.
Thanks.

File: authors/apps/articles/tests/test_filter_search_functionality.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        title = self.article.title
+        response = self.client.get(f"{self.url}?title={title}")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 1)
+        response = self.client.get(f"{self.url}?title=vdhfvsf")
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response.data), 0)
+
+    def test_user_can_filter_by_articles_by_tag(self):
+        """
```
Comment (fahdjamy) [inline authors/apps/articles/tests/test_filter_search_functionality.py:50]: I think the same thing might apply here. its test_user_can_filter_articles_by_tag.

**Review by fahdjamy — CHANGES_REQUESTED** [review]
Work looks good, precise and on point but add a little more comments in your classes plus generally make function names more descriptive and precise. like I commented on certain lines I've added on my review.
Otherwise, great Job done @oma0256.

File: authors/settings/base.py [hunk 3]
```diff
@@ -84,18 +85,27 @@
 # Password validation
 # https://docs.djangoproject.com/en/1.11/ref/settings/#auth-password-validators
 
+USER_ATTRIBUTE_SIMILARITY_VALIDATOR = \
```
Comment (missvicki) [inline authors/settings/base.py:88]: Thanks for the good work. If you don't mind me asking, what is the functionality of the lines below that were added in this file

Comment (oma0256) [reply authors/settings/base.py:88]: @missvicki I touched this file so they were pep8 issues I was just resolving those issues

Comment (missvicki) [reply authors/settings/base.py:88]: Alright, makes sense. You did a good job on this one

File: authors/settings/base.py [hunk 4]
```diff
@@ -84,18 +85,27 @@
...
-        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
+        'NAME': USER_ATTRIBUTE_SIMILARITY_VALIDATOR,
     },
     {
-        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
+        'NAME': MINIMUM_LENGTH_VALIDATOR,
     },
     {
-        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
+        'NAME': COMMON_PASSWORD_VALIDATOR,
```
Comment (pbkabali) [inline authors/settings/base.py:105]: @oma0256, good work here. Thanks. I see you have made some changes in the settings file. Do these two lines mean the same thing or some env variables need to be set in the env file?

Comment (oma0256) [reply authors/settings/base.py:105]: @pbkabali Yea they do mean the same thing I was just trying to avoid pep8 issues

Comment (pbkabali) [reply authors/settings/base.py:105]: OK cool. Thanks

**Review by fahdjamy — APPROVED** [review]
Thanks for implementing the requested changes.
LGTM right now.

**Review by pbkabali — APPROVED** [review]

**Comment by malep2007** [comment]
@oma0256 , please resolve the merge conflicts

**Pull request merged**